Improve failure message for string assertions when checking for equality

vbreuss opened the pull request:
When asserting that two strings are equal, and one of the strings is longer than 8 characters or contains a NewLine, instead of simply stating at which index the first difference occurs, this difference is made visible using arrows, e.g.
```csharp
var subject = "this is a long text that differs in between";
var expected = "this was too short";
subject.Should().Be(expected, "because we use arrows now");
```
throws with message
```
Expected subject to be the same string because we use arrows now, but they differ at index 5:
        ↓ (actual)
  "this is a long text…"
  "this was too short"
        ↑ (expected)."
```

For multi-line strings, instead of only providing information about the index, the failure message also contains the line and column information, e.g. `…but they differ on line 5 and column 1 (index 93):`

For short single-line strings the behaviour remains unchanged!

This closes #2050.

## IMPORTANT 

* [ ] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [ ] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [x] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [ ] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [ ] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).
    * [ ] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

dennisdoomen commented on Src/FluentAssertions/Common/StringExtensions.cs:
```diff
@@ -150,4 +151,105 @@ public static int CountSubstring(this string str, string substring, StringCompar
+    /// Get the mismatch segment between <paramref name="expected"/> and <paramref name="subject"/> for long strings,
+    /// when they differ at index <paramref name="firstIndexOfMismatch"/>.
+    /// </summary>
+    public static string GetMismatchSegmentForLongStrings(this string subject, string expected, int firstIndexOfMismatch)
```
🔧 This group of methods are not designed to be reusable and therefor should remain with the `StringEqualityStrategy`

vbreuss replied:
Moved them in [0ea0cda](https://github.com/fluentassertions/fluentassertions/pull/2307/commits/0ea0cdab575b8956d1c9f1d298951ecec216de3c).

dennisdoomen commented on Src/FluentAssertions/Common/StringExtensions.cs:
```diff
@@ -150,4 +151,105 @@ public static int CountSubstring(this string str, string substring, StringCompar
+    {
+        int trimStart = CalculateSegmentStart(subject, firstIndexOfMismatch);
+
+        int whiteSpaceCount = (firstIndexOfMismatch - trimStart) + 3;
```
❓ I don't get what you are trying to do here. Where is the `3` coming from and why do you call this `whiteSpaceCount`. I'm pretty sure it contains other characters as well.

vbreuss replied:
It's the number of visible white-space before the arrow symbol. The `3` comes from the length of the prefix `  "` that's before every text and the optional increment considers the ellipsis that is included.
I renamed it to `whiteSpaceCountBeforeArrow` and replaced the `3` with `prefix.Length`.

dennisdoomen commented on Src/FluentAssertions/Common/StringExtensions.cs:
```diff
@@ -150,4 +151,105 @@ public static int CountSubstring(this string str, string substring, StringCompar
+    /// Calculates the start index of the visible segment from <paramref name="value"/> when highlighting the difference at <paramref name="index"/>.<br />
+    /// Either keep the last 10 characters before <paramref name="index"/> or a word begin (separated by whitespace) between 15 and 5 characters before <paramref name="index"/>.
+    /// </summary>
+    private static int CalculateSegmentStart(string value, int index)
```
🤔 I think that naming this something like `GetStartIndexOfPhraseToShowBeforeTheMismatchingIndex` covers its purpose better. I had to use the debugger to understand this. 
🔧 Rename `index` to `indexOfFirstMismatch`

vbreuss replied:
Renamed to `GetStartIndexOfPhraseToShowBeforeTheMismatchingIndex` in [9becb00](https://github.com/fluentassertions/fluentassertions/pull/2307/commits/9becb00b59cf8e3fcb7ed2bfdec5b05582b2b44b).

dennisdoomen commented on Src/FluentAssertions/Common/StringExtensions.cs:
```diff
@@ -150,4 +151,105 @@ public static int CountSubstring(this string str, string substring, StringCompar
+    /// </summary>
+    private static int CalculateSegmentStart(string value, int index)
+    {
+        if (index <= 10)
```
🔧 Use a `const` for this.

vbreuss replied:
Used `const` values for `defaultCharactersToKeep`, `minCharactersToKeep`, `maxCharactersToKeep` and `lengthOfWhitespace` in [9becb00](https://github.com/fluentassertions/fluentassertions/pull/2307/commits/9becb00b59cf8e3fcb7ed2bfdec5b05582b2b44b).

dennisdoomen commented on Src/FluentAssertions/Common/StringExtensions.cs:
```diff
@@ -150,4 +151,105 @@ public static int CountSubstring(this string str, string substring, StringCompar
+            return 0;
+        }
+
+        var wordSearchBegin = Math.Max(index - 16, 0);
```
🔧 You refer to `15` in the docs, so use a `const` to show how you got to `16.
🔧 Rename `wordSearchBegin` to something like `indexToStartSearchingForWordBoundary`

vbreuss replied:
Used `const` values for `defaultLength`, `minLength`, `maxLength` and `lengthOfWhitespace` in [9becb00](https://github.com/fluentassertions/fluentassertions/pull/2307/commits/9becb00b59cf8e3fcb7ed2bfdec5b05582b2b44b).

dennisdoomen commented on Src/FluentAssertions/Common/StringExtensions.cs:
```diff
@@ -150,4 +151,105 @@ public static int CountSubstring(this string str, string substring, StringCompar
+
+        var wordSearchBegin = Math.Max(index - 16, 0);
+
+        var wordIndex = value.Substring(wordSearchBegin, 8)
```
🤔 I think this is the `indexOfWordBoundary` or `indexOfWhitespace`

vbreuss replied:
Renamed to `indexOfWordBoundary`

dennisdoomen commented on Src/FluentAssertions/Common/StringExtensions.cs:
```diff
@@ -150,4 +151,105 @@ public static int CountSubstring(this string str, string substring, StringCompar
+    }
+
+    /// <summary>
+    /// Calculates how many characters to keep in <paramref name="value"/>.<br />
```
🔧 You don't need to use `<br/?>` if you put the second line of text in a `<remarks>` element.

dennisdoomen commented on Src/FluentAssertions/Common/StringExtensions.cs:
```diff
@@ -150,4 +151,105 @@ public static int CountSubstring(this string str, string substring, StringCompar
+    /// Calculates how many characters to keep in <paramref name="value"/>.<br />
+    /// If a word end is found between 15 and 25 characters, use this word end, otherwise keep 20 characters.
+    /// </summary>
+    private static int CalculateSegmentLength(string value)
```
🔧 Rename this to something more functional like `GetWordBoundaryOrDefaultLength` and return the actual substring. It'll make the invoking code easier to understand.
🔧 Replace the magic numbers so it's clearer how they relate to what you put in the comments.

vbreuss replied:
Renamed it to `GetLengthOfPhraseToShowOrDefaultLength` and replaced the magic numbers

dennisdoomen commented on Src/FluentAssertions/Common/StringExtensions.cs:
```diff
@@ -150,4 +151,105 @@ public static int CountSubstring(this string str, string substring, StringCompar
+        }
+
+        var visibleText = subject.Substring(trimStart, firstIndexOfMismatch - trimStart);
+        whiteSpaceCount += visibleText.Count(c => c is '\r' or '\n');
```
🤔 Does this work correctly on all operating systems? I guess that's why we have `Environment.NewLine`, right?

vbreuss replied:
As I have to replace `\r` with `\\r` and `\n` with `\\n` in the output (so that newlines are visible), I have to consider this when calculating where the arrows should point to.

dennisdoomen commented on Src/FluentAssertions/Common/StringExtensions.cs:
```diff
@@ -150,4 +151,105 @@ public static int CountSubstring(this string str, string substring, StringCompar
+
+        if (trimStart > 0)
+        {
+            sb.Append('\u2026');
```
🔧 Replace with a `const`. I had to look it up to understand what it was.

vbreuss replied:
Used `const` values for `ellipsis` in [9becb00](https://github.com/fluentassertions/fluentassertions/pull/2307/commits/9becb00b59cf8e3fcb7ed2bfdec5b05582b2b44b).

dennisdoomen commented on Src/FluentAssertions/Common/StringExtensions.cs:
```diff
@@ -150,4 +151,105 @@ public static int CountSubstring(this string str, string substring, StringCompar
+        return sb.ToString();
+    }
+
+    private static StringBuilder AppendVisibleText(this StringBuilder sb, string subject, int trimStart)
```
❓ What does this method try to do? It seems to take a substring and optionally adds a triple elipsis under certain circumstances.

vbreuss replied:
I renamed it to `AppendPrefixAndEscapedPhraseToShowWithEllipsisAndSuffix` and added a description in [9becb00](https://github.com/fluentassertions/fluentassertions/pull/2307/commits/9becb00b59cf8e3fcb7ed2bfdec5b05582b2b44b).

It's a helper method to print `  "{optional leading ellipsis}{escaped visible text phrase}{optional trailing ellipsis}"\n` for the actual and the expected string.

dennisdoomen commented on Src/FluentAssertions/Common/StringExtensions.cs:
```diff
@@ -150,4 +151,105 @@ public static int CountSubstring(this string str, string substring, StringCompar
+
+        var sb = new StringBuilder();
+
+        sb.Append(' ', whiteSpaceCount).AppendLine("\u2193 (actual)")
```
🔧 Use `const` for all usage of unicode characters.

vbreuss replied:
Used `const` values for `arrowDown` and `arrowUp` in [9becb00](https://github.com/fluentassertions/fluentassertions/pull/2307/commits/9becb00b59cf8e3fcb7ed2bfdec5b05582b2b44b).

dennisdoomen commented on Src/FluentAssertions/Primitives/StringEqualityStrategy.cs:
```diff
@@ -57,19 +58,47 @@ private string GetMismatchSegmentForStringsOfDifferentLengths(string subject, st
+                return;
+            }
+
+            string lineInfo = $"at index {indexOfMismatch}";
```
🤔 I think naming this `locationDescription` or something like that will help explain the purpose

vbreuss replied:
Renamed to `locationDescription` in [9becb00](https://github.com/fluentassertions/fluentassertions/pull/2307/commits/9becb00b59cf8e3fcb7ed2bfdec5b05582b2b44b).

dennisdoomen commented on Src/FluentAssertions/Primitives/StringEqualityStrategy.cs:
```diff
@@ -57,19 +58,47 @@ private string GetMismatchSegmentForStringsOfDifferentLengths(string subject, st
+
+            if (lineNumber > 0)
+            {
+                var lastLineIndex = matchingString.LastIndexOf('\n');
```
🤔 Did you mean `lastLineNumber`?

vbreuss replied:
No, it is the index of the last newline char before the `indexOfMismatch` and this allows me to calculate the `column` in the line.
I renamed it to `indexOfLastNewlineBeforeMismatch`.

dennisdoomen reviewed (changes requested):
That's a pretty impressive improvement. I had some quite some issues understanding the algorithm, so I've gone through a pretty thorough review.

ITaluone commented on Tests/FluentAssertions.Specs/Primitives/StringAssertionSpecs.Be.cs:
```diff
@@ -231,11 +231,71 @@ public void When_text_is_longer_than_8_characters_only_add_ellipsis_where_applic
                 + "  \"this was too short\"" + Environment.NewLine
                 + "        ↑ (expected).");
         }
 
```
What about adding the expected message as raw string?

like:
```csharp

string msg = """
             Expected subject to be the same string because we use arrows now, but they differ at index 5:
                    ↓ (actual)
              "this is a long text that…"
              "this was too short"
                    ↑ (expected).
             """;
```

This allows to read quickly through the tests and see where the pointers are pointing to. Now I have to stop an "think the `\` away" in my mind to get the actual position of the arrows.

vbreuss replied:
I just did this with [6cbf802](https://github.com/fluentassertions/fluentassertions/pull/2307/commits/6cbf80244606b611ac14bfb1db1b1fc746b219d3) :-)

ITaluone replied:
Oh.. I see.. 🎉

vbreuss commented:
> That's a pretty impressive improvement. I had some quite some issues understanding the algorithm, so I've gone through a pretty thorough review.

@dennisdoomen. Thanks for the extensive review. It's really appreciated!

I implemented your requested changes and answered your questions. I hope I didn't overlook anything :-)
I also added this change to the [`releases.md`](https://github.com/vbreuss/fluentassertions/blob/topic/improve-string-messages/docs/_pages/releases.md#improvements)

jnyrup commented on Tests/FluentAssertions.Specs/FluentAssertions.Specs.csproj:
```diff
@@ -7,6 +7,7 @@
     <IsPackable>false</IsPackable>
     <NoWarn>$(NoWarn),IDE0052,1573,1591,1712,CS8002</NoWarn>
     <DebugType>full</DebugType>
+    <langVersion>latest</langVersion>
```
We already specify `<LangVersion>11.0</LangVersion>`
https://github.com/fluentassertions/fluentassertions/blob/a8a04fe96658a53c6747bfc1ea485820b1edbcb7/Directory.Build.props#L3C5-L3C5

vbreuss replied:
Removed `<langVersion>latest</langVersion> in [110ed76](https://github.com/fluentassertions/fluentassertions/pull/2307/commits/110ed76ff676dace0b57314fc71bd65f28abf5fe).

jnyrup reviewed (commented):
This is really great!

I haven't gone through all the relational operators to check if they are correct.
I'll try to run mutation tests on `StringEqualityStrategy.cs` sometime this week to help verifying this. If you want to try mutation tests before I get to it, there are some pointers in #1068 and https://github.com/jnyrup/fluentassertions/pull/6

jnyrup commented:
> > I'll try to run mutation tests on StringEqualityStrategy.cs sometime this week to help verifying this.
> 
> Shall we add that to our pipeline?

That's my plan.
It takes quite a while to do a full run, so my current idea is to only consider the changed files in PRs and then do a scheduled full run e.g. once a week.

vbreuss commented:
> I'll try to run mutation tests on `StringEqualityStrategy.cs` sometime this week to help verifying this. If you want to try mutation tests before I get to it, there are some pointers in #1068 and [jnyrup#6](https://github.com/jnyrup/fluentassertions/pull/6)

I also use Stryker in [my project](https://github.com/Testably/Testably.Abstractions) and just ran it against fluentassertions

### Mutation Testing Summary
| File                                                                 | Score  | Killed | Survived | Timeout | No Coverage | Ignored | Compile Errors | Total Detected | Total Undetected | Total Mutants |
| -------------------------------------------------------------------- | ------ | ------ | -------- | ------- | ----------- | ------- | -------------- | -------------- | ---------------- | ------------- |
| Primitives\\StringEqualityStrategy.cs                                | 97,12% | 101    | 3        | 0       | 0           | 21      | 8              | 101            | 3                | 133           |

I will add tests for the 3 survived mutants together with the changes for your review  :-)

vbreuss commented on Src/FluentAssertions/Primitives/StringEqualityStrategy.cs:
```diff
@@ -84,4 +115,122 @@ public string ExpectationDescription
+        const int maxLength = 25;
+        const int lengthOfWhitespace = 1;
+
+        var indexOfWordBoundary = value[..Math.Min(maxLength + lengthOfWhitespace, value.Length)]
+            .LastIndexOf(' ');
```
It's because in this overload the parameter is the 0-based [starting position of the search](https://learn.microsoft.com/en-us/dotnet/api/system.string.lastindexof?view=net-6.0#system-string-lastindexof(system-char-system-int32)).

jnyrup replied:
I did figure out why `-1` was necessary for the statement to be equivalent.
It's whether that was desirable I didn't get to look into.

vbreuss replied:
Fixed in [110ed76](https://github.com/fluentassertions/fluentassertions/pull/2307/commits/110ed76ff676dace0b57314fc71bd65f28abf5fe).

vbreuss commented on Src/FluentAssertions/Primitives/StringEqualityStrategy.cs:
```diff
@@ -84,4 +115,122 @@ public string ExpectationDescription
+            whiteSpaceCountBeforeArrow++;
+        }
+
+        var visibleText = subject.Substring(trimStart, firstIndexOfMismatch - trimStart);
```
Fixed in [110ed76](https://github.com/fluentassertions/fluentassertions/pull/2307/commits/110ed76ff676dace0b57314fc71bd65f28abf5fe).

vbreuss commented on Src/FluentAssertions/Primitives/StringEqualityStrategy.cs:
```diff
@@ -84,4 +115,122 @@ public string ExpectationDescription
+
+        var indexOfWordBoundary = value
+            .Substring(indexToStartSearchingForWordBoundary, (maxCharactersToKeep - minCharactersToKeep) + lengthOfWhitespace)
+            .IndexOf(' ', StringComparison.OrdinalIgnoreCase);
```
Changed it in [110ed76](https://github.com/fluentassertions/fluentassertions/pull/2307/commits/110ed76ff676dace0b57314fc71bd65f28abf5fe), but please note, that I then have to subtract the `startIndex` afterwards.

vbreuss commented:
@jnyrup 
I implemented your review comments and added the missing tests for the surviving mutants.

### Mutation Testing Summary

| File                                                                 | Score   | Killed | Survived | Timeout | No Coverage | Ignored | Compile Errors | Total Detected | Total Undetected | Total Mutants |
| -------------------------------------------------------------------- | ------- | ------ | -------- | ------- | ----------- | ------- | -------------- | -------------- | ---------------- | ------------- |
| Primitives\\StringEqualityStrategy.cs                                | 100,00% | 103    | 0        | 0       | 0           | 20      | 8              | 103            | 0                | 131           |

dennisdoomen commented on Tests/FluentAssertions.Specs/Primitives/StringAssertionSpecs.Be.cs:
```diff
@@ -184,8 +189,169 @@ public void When_two_strings_differ_and_one_of_them_is_multiline_it_should_displ
+        }
+
+        [Fact]
+        public void When_empty_string_is_compared_with_long_text_should_differ_in_length()
```
```suggestion
        public void An_empty_string_is_always_shorter_than_a_long_text()
```

dennisdoomen commented on Tests/FluentAssertions.Specs/Primitives/StringAssertionSpecs.Be.cs:
```diff
@@ -184,8 +189,169 @@ public void When_two_strings_differ_and_one_of_them_is_multiline_it_should_displ
+        }
+
+        [Fact]
+        public void When_long_text_has_first_mismatch_in_first_10_characters_should_include_whole_start_of_the_text()
```
```suggestion
        public void A_mismatch_below_index_11_includes_all_text_preceding_the_index_in_the_failure()
```

dennisdoomen commented on Tests/FluentAssertions.Specs/Primitives/StringAssertionSpecs.Be.cs:
```diff
@@ -184,8 +189,169 @@ public void When_two_strings_differ_and_one_of_them_is_multiline_it_should_displ
+        }
+
+        [Fact]
+        public void When_text_has_many_lines_failure_message_should_look_informative()
```
```suggestion
        public void Mismatches_in_multiline_text_includes_the_line_number()
```

dennisdoomen reviewed (approved):
As I said before, great work. The new comments and improved naming really helps (even though it is still quite an algorithm). 

Since we've tried to get rid of words like `when` and `should` in test names and focus more on the fact that such a test covers, I've suggested some improvements. Up to you if you want to include them now. Otherwise, I'll do it at a later point.

vbreuss commented:
@dennisdoomen :
Thanks for the hint, I was not aware of the naming convention for tests.
I adapted the test names (mostly) according to your suggestions :-)

ITaluone commented:
Pretty cool and useful enhancement!

Thank you 🎉

thomOrbelius commented:
Thank you for this improvement!
